swarm_dial.go

```diff
@@ -290,6 +290,9 @@ func (s *Swarm) dial(ctx context.Context, p peer.ID) (*Conn, error) {
 		that we previously had (halting a dial when we run out of addrs)
 	*/
 	goodAddrs := s.filterKnownUndialables(s.peers.Addrs(p))
+	if len(goodAddrs) == 0 {
+		return nil, errors.New("no good addresses")
```

Comment:
IMO, use exported error value instead of inline errors.

Reply:
Sure, if they have value to consumers. Here I'm simply providing a better textual error than was previously available. I don't want to commit prematurely to any meaning.

Reply:
Allowing consumers to distinguish when a dial failed is valuable. Precisely by exporting this error you won't be committing to anything because you can change the text anytime without breaking consumers. If you don't export it, you set yourself up for users doing textual comparisons on the error message if they wan't to distinguish this situation.

Reply:
That's fair enough, but is it even an actionable error?

Reply:
Yes, it is.

Reply:
@raulk How? You try to dial a peer and can't because you have no addresses for them. You don't then add them and try again, I don't think anybody does that. If you get a dial error because you had no good addresses, you should have already added all the addresses you can think of.